OpenRA.Game/Map/ActorInitializer.cs

```diff
@@ -53,59 +65,175 @@ public ActorInitializer(Actor actor, TypeDictionary dict)
 			return init;
 		}
 
-		public U GetValue<T, U>(TraitInfo info) where T : IActorInit<U>
+		public U GetValue<T, U>(TraitInfo info) where T : ValueActorInit<U>
 		{
 			return Get<T>(info).Value;
 		}
 
-		public U GetValue<T, U>(TraitInfo info, U fallback) where T : IActorInit<U>
+		public U GetValue<T, U>(TraitInfo info, U fallback) where T : ValueActorInit<U>
 		{
 			var init = GetOrDefault<T>(info);
 			return init != null ? init.Value : fallback;
 		}
 
-		public bool Contains<T>(TraitInfo info) where T : IActorInit { return GetOrDefault<T>(info) != null; }
+		public bool Contains<T>(TraitInfo info) where T : ActorInit { return GetOrDefault<T>(info) != null; }
 	}
 
-	public interface IActorInit { }
+	/*
+	 * Things to be aware of when writing ActorInits:
+	 *
+	 * - ActorReference and ActorGlobal can dynamically create objects without calling a constructor.
+	 *   The object will be allocated directly then the best matching Initialize() method will be called to set valid state.
+	 * - ActorReference will always attempt to call Initialize(MiniYaml). ActorGlobal will use whichever one it first
+	 *   finds with an argument type that matches the given LuaValue.
+	 * - Most ActorInits will want to inherit either ValueActorInit<T> or CompositeActorInit which hide the low-level plumbing.
+	 * - Inits that reference actors should use ActorInitActorReference which allows actors to be referenced by name in map.yaml
+	 */
+	public abstract class ActorInit
+	{
+		[FieldLoader.Ignore]
+		public readonly string InstanceName;
+
+		protected ActorInit(string instanceName)
+		{
+			InstanceName = instanceName;
+		}
+
+		protected ActorInit() { }
 
-	public interface IActorInit<T> : IActorInit
+		public abstract MiniYaml Save();
+	}
+
+	public abstract class ValueActorInit<T> : ActorInit
 	{
-		T Value { get; }
+		protected readonly T value;
+
+		protected ValueActorInit(TraitInfo info, T value)
+			: base(info.InstanceName) { this.value = value; }
+
+		protected ValueActorInit(T value) { this.value = value; }
+
+		public virtual T Value { get { return value; } }
+
+		public virtual void Initialize(MiniYaml yaml)
+		{
+			var valueType = typeof(T).IsEnum ? Enum.GetUnderlyingType(typeof(T)) : typeof(T);
+			Initialize((T)FieldLoader.GetValue("value", valueType, yaml.Value));
+		}
+
+		public virtual void Initialize(T value)
+		{
+			var field = GetType().GetField("value", BindingFlags.NonPublic | BindingFlags.Instance);
+			if (field != null)
+				field.SetValue(this, value);
+		}
+
+		public override MiniYaml Save()
+		{
+			return new MiniYaml(FieldSaver.FormatValue(value));
+		}
 	}
 
-	public class LocationInit : IActorInit<CPos>
+	public abstract class CompositeActorInit : ActorInit
 	{
-		[FieldFromYamlKey]
-		readonly CPos value = CPos.Zero;
+		protected CompositeActorInit(TraitInfo info)
+			: base(info.InstanceName) { }
```

Comment:
This crashes for me when killing a unit that is on the chronoshift countdown since `info` is null.